diff --git a/src/OpenFOAM/db/IOstreams/Pstreams/UPstream.H b/src/OpenFOAM/db/IOstreams/Pstreams/UPstream.H
index b9f08e0b..269fddb7 100644
--- a/src/OpenFOAM/db/IOstreams/Pstreams/UPstream.H
+++ b/src/OpenFOAM/db/IOstreams/Pstreams/UPstream.H
@@ -369,6 +369,7 @@ public:
         //- Initialisation function called from main
         //  Spawns slave processes and initialises inter-communication
         static bool init(int& argc, char**& argv, const bool needsThread);
+        static bool init(void* comm, const bool needsThread);
 
         // Non-blocking comms
 
diff --git a/src/OpenFOAM/global/argList/argList.C b/src/OpenFOAM/global/argList/argList.C
index 6f3d5b57..be434a27 100644
--- a/src/OpenFOAM/global/argList/argList.C
+++ b/src/OpenFOAM/global/argList/argList.C
@@ -574,6 +574,154 @@ Foam::argList::argList
 }
 
 
+Foam::argList::argList
+(
+    int& argc,
+    char**& argv,
+    void* comm,
+    bool checkArgs,
+    bool checkOpts,
+    const bool initialise
+)
+:
+    args_(argc),
+    options_(argc)
+{
+    // Pre-load any libraries
+    {
+        const string libsString(getEnv("FOAM_LIBS"));
+        if (!libsString.empty())
+        {
+            libs.open(fileNameList((IStringStream(libsString))()));
+        }
+
+        for (int argi=0; argi<argc; argi++)
+        {
+            if (argv[argi][0] == '-')
+            {
+                const char *optionName = &argv[argi][1];
+                if (string(optionName) == "libs")
+                {
+                    libs.open(fileNameList((IStringStream(argv[argi+1]))()));
+                    break;
+                }
+            }
+        }
+    }
+
+    // Check for fileHandler
+    word handlerType(getEnv("FOAM_FILEHANDLER"));
+    for (int argi=0; argi<argc; argi++)
+    {
+        if (argv[argi][0] == '-')
+        {
+            const char *optionName = &argv[argi][1];
+            if (string(optionName) == "fileHandler")
+            {
+                handlerType = argv[argi+1];
+                break;
+            }
+        }
+    }
+    if (handlerType.empty())
+    {
+        handlerType = fileOperation::defaultFileHandler;
+    }
+
+    // Detect any parallel options
+    bool needsThread = fileOperations::fileOperationInitialise::New
+    (
+        handlerType,
+        argc,
+        argv
+    )().needsThreading();
+
+
+    // Check if this run is a parallel run by searching for any parallel option
+    // If found call runPar which might filter argv
+    for (int argi=0; argi<argc; argi++)
+    {
+        if (argv[argi][0] == '-')
+        {
+            const char *optionName = &argv[argi][1];
+
+            if (validParOptions.found(optionName))
+            {
+                parRunControl_.runPar(comm, needsThread);
+                break;
+            }
+        }
+    }
+
+    // Convert argv -> args_ and capture ( ... ) lists
+    // for normal arguments and for options
+    regroupArgv(argc, argv);
+
+    // Get executable name
+    args_[0]    = fileName(argv[0]);
+    executable_ = fileName(argv[0]).name();
+
+    // Check arguments and options, we already have argv[0]
+    int nArgs = 1;
+    argListStr_ = args_[0];
+
+    for (int argi=1; argi<args_.size(); argi++)
+    {
+        argListStr_ += ' ';
+        argListStr_ += args_[argi];
+
+        if (args_[argi][0] == '-')
+        {
+            const char *optionName = &args_[argi][1];
+
+            if
+            (
+                (
+                    validOptions.found(optionName)
+                 && !validOptions[optionName].empty()
+                )
+             || (
+                    validParOptions.found(optionName)
+                 && !validParOptions[optionName].empty()
+                )
+            )
+            {
+                ++argi;
+                if (argi >= args_.size())
+                {
+                    FatalError
+                        <<"Option '-" << optionName
+                        << "' requires an argument" << endl;
+                    printUsage();
+                    FatalError.exit();
+                }
+
+                argListStr_ += ' ';
+                argListStr_ += args_[argi];
+                options_.insert(optionName, args_[argi]);
+            }
+            else
+            {
+                options_.insert(optionName, "");
+            }
+        }
+        else
+        {
+            if (nArgs != argi)
+            {
+                args_[nArgs] = args_[argi];
+            }
+            ++nArgs;
+        }
+    }
+
+    args_.setSize(nArgs);
+
+    parse(checkArgs, checkOpts, initialise);
+}
+
+
+
 Foam::argList::argList
 (
     const argList& args,
diff --git a/src/OpenFOAM/global/argList/argList.H b/src/OpenFOAM/global/argList/argList.H
index 7b504a08..a09833d2 100644
--- a/src/OpenFOAM/global/argList/argList.H
+++ b/src/OpenFOAM/global/argList/argList.H
@@ -201,6 +201,16 @@ public:
             bool initialise = true
         );
 
+        argList
+        (
+            int& argc,
+            char**& argv,
+            void* comm,
+            bool checkArgs = true,
+            bool checkOpts = true,
+            bool initialise = true
+        );
+
         //- Construct copy with new options
         argList
         (
diff --git a/src/OpenFOAM/global/argList/parRun.H b/src/OpenFOAM/global/argList/parRun.H
index e8de3d05..6ea92e55 100644
--- a/src/OpenFOAM/global/argList/parRun.H
+++ b/src/OpenFOAM/global/argList/parRun.H
@@ -64,6 +64,17 @@ public:
         }
     }
 
+    void runPar(void* external_comm, const bool needsThread)
+    {
+        RunPar = true;
+
+        if (!Pstream::init(external_comm, needsThread))
+        {
+            Info<< "Failed to start parallel run" << endl;
+            Pstream::exit(1);
+        }
+    }
+
     void runPar(int& argc, char**& argv, const bool needsThread)
     {
         RunPar = true;
diff --git a/src/Pstream/dummy/UPstream.C b/src/Pstream/dummy/UPstream.C
index f0553bae..257a1082 100644
--- a/src/Pstream/dummy/UPstream.C
+++ b/src/Pstream/dummy/UPstream.C
@@ -43,6 +43,17 @@ bool Foam::UPstream::init(int& argc, char**& argv, const bool needsThread)
 }
 
 
+bool Foam::UPstream::init(void*, const bool)
+{
+    FatalErrorInFunction
+        << "Trying to use the dummy Pstream library." << nl
+        << "This dummy library cannot be used in parallel mode"
+        << Foam::exit(FatalError);
+
+    return false;
+}
+
+
 void Foam::UPstream::exit(int errnum)
 {
     NotImplemented;
diff --git a/src/Pstream/mpi/UPstream.C b/src/Pstream/mpi/UPstream.C
index 88c34ea8..93b7459e 100644
--- a/src/Pstream/mpi/UPstream.C
+++ b/src/Pstream/mpi/UPstream.C
@@ -50,6 +50,8 @@ License
 // valid parallel options vary between implementations, but flag common ones.
 // if they are not removed by MPI_Init(), the subsequent argument processing
 // will notice that they are wrong
+static bool _external_mpi = false;
+
 void Foam::UPstream::addValidParOptions(HashTable<string>& validParOptions)
 {
     validParOptions.insert("np", "");
@@ -61,6 +63,88 @@ void Foam::UPstream::addValidParOptions(HashTable<string>& validParOptions)
 }
 
 
+
+bool Foam::UPstream::init(void* comm_ptr, const bool needsThread)
+{
+    _external_mpi = true;
+    MPI_Comm comm = *((MPI_Comm*)comm_ptr);
+
+    int flag;
+    MPI_Initialized(&flag);
+    if (!flag)
+    {
+        FatalErrorInFunction
+            << "MPI not initialised: call init(int& argc, char**& argv, const bool needsThread)" << nl
+            << Foam::abort(Foam::FatalError);
+    }
+
+    int provided_thread_support = MPI_THREAD_SINGLE;
+    if (needsThread)
+    {
+        MPI_Query_thread(&provided_thread_support);
+        if (provided_thread_support != MPI_THREAD_MULTIPLE)
+        {
+            FatalErrorInFunction
+                << "MPI implementation does not provide required level of threading support (MPI_THREAD_MULTIPLE)"
+                << Foam::abort(Foam::FatalError);
+        }
+    }
+
+    int myGlobalRank;
+    MPI_Comm_rank(comm, &myGlobalRank);
+    MPI_Comm_split
+    (
+        comm,
+        1,
+        myGlobalRank,
+        &PstreamGlobals::MPI_COMM_FOAM
+    );
+
+    int numprocs;
+    MPI_Comm_size(PstreamGlobals::MPI_COMM_FOAM, &numprocs);
+    int myRank;
+    MPI_Comm_rank(PstreamGlobals::MPI_COMM_FOAM, &myRank);
+
+    if (debug)
+    {
+        Pout<< "UPstream::init : initialised with numProcs:" << numprocs
+            << " myRank:" << myRank << endl;
+    }
+
+    if (numprocs <= 1)
+    {
+        FatalErrorInFunction
+            << "bool IPstream::init(int& argc, char**& argv) : "
+               "attempt to run parallel on 1 processor"
+            << Foam::abort(FatalError);
+    }
+
+    setParRun(numprocs, provided_thread_support == MPI_THREAD_MULTIPLE);
+
+    #ifndef SGIMPI
+    string bufferSizeName = getEnv("MPI_BUFFER_SIZE");
+
+    if (bufferSizeName.size())
+    {
+        int bufferSize = atoi(bufferSizeName.c_str());
+
+        if (bufferSize)
+        {
+            MPI_Buffer_attach(new char[bufferSize], bufferSize);
+        }
+    }
+    else
+    {
+        FatalErrorInFunction
+            << "UPstream::init(int& argc, char**& argv) : "
+            << "environment variable MPI_BUFFER_SIZE not defined"
+            << Foam::abort(FatalError);
+    }
+    #endif
+
+    return true;
+}
+
 bool Foam::UPstream::init(int& argc, char**& argv, const bool needsThread)
 {
     // MPI_Init(&argc, &argv);
@@ -185,8 +269,11 @@ void Foam::UPstream::exit(int errnum)
 
     if (errnum == 0)
     {
-        MPI_Finalize();
-        ::exit(errnum);
+        if (!_external_mpi)
+        {
+            MPI_Finalize();
+            ::exit(errnum);
+        }
     }
     else
     {
